bpf-map-pressure-exporter/src/fetcher.go

```diff
@@ -0,0 +1,181 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/cybozu-go/log"
+)
+
+type IBpfMapPressureFetcher interface {
+	GetMetrics() []bpfMapPressure
+	Start(context.Context)
+}
+
+type bpfMapPressureFetcher struct {
+	mapNameStrings []string
+	fetchInterval  time.Duration
+	mutex          sync.RWMutex
+	metrics        []bpfMapPressure
+}
+
+type bpfMapPressure struct {
+	mapId       uint32
+	mapName     string
+	mapPressure float64
+}
+
+func newFetcher(mapNameStrings []string, interval time.Duration) *bpfMapPressureFetcher {
+	return &bpfMapPressureFetcher{
+		mapNameStrings: mapNameStrings,
+		fetchInterval:  interval,
+	}
+}
+
+func (f *bpfMapPressureFetcher) Start(ctx context.Context) {
+	f.update()
+
+	ticker := time.NewTicker(f.fetchInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			f.update()
+		}
+	}
+}
+
+func (f *bpfMapPressureFetcher) update() {
+	metrics := []bpfMapPressure{}
+
+	defer func() {
+		f.mutex.Lock()
+		defer f.mutex.Unlock()
+		f.metrics = metrics
+	}()
+
+	metrics = f.fetch()
+}
+
+func (f *bpfMapPressureFetcher) GetMetrics() []bpfMapPressure {
+	f.mutex.RLock()
+	defer f.mutex.RUnlock()
+	return f.metrics
+}
+
+func createBuffer(structSize int, elemNum int) (typ reflect.Type, buf any) {
+	fields := []reflect.StructField{}
+	for i := 0; i < structSize; i++ {
+		fields = append(fields, reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: reflect.TypeOf(uint8(0)),
+		})
+	}
+	typ = reflect.StructOf(fields)
+	buf = reflect.MakeSlice(reflect.SliceOf(typ), elemNum, elemNum).Interface()
+	return typ, buf
+}
+
+func (f *bpfMapPressureFetcher) fetch() []bpfMapPressure {
+	results := []bpfMapPressure{}
+	var id ebpf.MapID = 0
+	var err error
+	const chunk uint32 = 4096
+	for {
+		id, err = ebpf.MapGetNextID(id)
+		if errors.Is(err, os.ErrNotExist) {
+			break
+		}
+		if err != nil {
+			_ = logger.Warn("failed to get next map id", map[string]interface{}{
+				"id":        id,
+				log.FnError: err,
+			})
+			return results
+		}
+		m, err := ebpf.NewMapFromID(id)
+		if err != nil {
+			_ = logger.Warn("failed to get map", map[string]interface{}{
+				"id":        id,
+				log.FnError: err,
+			})
+			return results
+		}
+		info, err := m.Info()
+		if err != nil {
+			_ = logger.Warn("failed to get map info", map[string]interface{}{
+				"id":        id,
+				log.FnError: err,
+			})
+			return results
+		}
+		for _, str := range f.mapNameStrings {
+			if strings.Contains(info.Name, str) {
```

Comment:
```suggestion
			if !strings.Contains(info.Name, str) {
				continue
			}
			...
```

nit: I'd write like these :)